src/entities/nestedExit/getNestedExitCalls.ts

```diff
@@ -0,0 +1,193 @@
+import { Token } from '../token';
+import { NestedExitInput, NestedExitCall } from './types';
+import { NestedPool, NestedPoolState } from '../types';
+import { TokenAmount } from '../tokenAmount';
+import { Address } from '../../types';
+
+export const getNestedExitCalls = (
+    {
+        bptAmountIn,
+        chainId,
+        accountAddress,
+        tokenOut,
+        useNativeAssetAsWrappedAmountOut = false,
+        toInternalBalance = false,
+    }: NestedExitInput,
+    { pools }: NestedPoolState,
+): { bptAmountIn: TokenAmount; calls: NestedExitCall[] } => {
+    const isProportional = tokenOut === undefined;
+    let poolsTopDown: NestedPool[];
+    let calls: NestedExitCall[];
+
+    if (isProportional) {
+        /**
+         * Overall logic to build sequence of proportional exit calls:
+         * 1. Go from top pool to bottom filling out input amounts and output refs
+         * 2. Inputs will be bptAmountIn provided or output of the previous level
+         * 3. Output at bottom level is the amountsOut
+         */
+
+        // sort pools by descending level
+        poolsTopDown = pools.sort((a, b) => b.level - a.level);
+
+        calls = getProportionalExitCalls(
+            poolsTopDown,
+            chainId,
+            useNativeAssetAsWrappedAmountOut,
+            accountAddress,
+            bptAmountIn,
+            toInternalBalance,
+        );
+    } else {
+        /**
+         * Overall logic to build sequence of single token exit calls:
+         * 1. Go BOTTOM-UP building exit path to tokenOut
+         * 2. Go through exit path filling out input amounts and output refs
+         * 3. Inputs will be bptAmountIn provided or output of the previous level
+         * 4. Output at bottom level is the amountOut
+         */
+
+        // sort pools by descending level
+        poolsTopDown = pools.sort((a, b) => b.level - a.level);
+        const topPool = poolsTopDown[0];
+
+        // Go BOTTOM-UP building exit path to tokenOut
+        const exitPath: NestedPool[] = [];
+        let tokenOutByLevel = tokenOut;
+        while (tokenOutByLevel !== topPool.address) {
+            const currentPool = poolsTopDown.find(
+                (p) =>
+                    p.address !== tokenOutByLevel && // prevents pools with BPT as token to be picked up incorrectly
```

Comment:
Could you give me an example case of this when you get a chance?

Reply:
Sure! The pool being used on the tests (WETH/3-POOL) has a case where this happens.
When exiting to DAI for example, the first iteration will pick 3-POOL as the "bottom" pool and update tokenOutByLevel to 3-POOL-BPT. Since 3-POOL-BPT is contained on both WETH/3-POOL and 3-POOL itself, simply checking if the pool contains that token could result in the bottom pool being picked up again.
That's why I have to filter out pools that have tokenOutByLevel as it's own address.
Let me know if that's still not clear 😅

Reply:
Thanks for clarifying. Would you add similar as a code comment. I think it'll help if we ever come back to this code in future.